sor/src/main/java/com/bazaarvoice/emodb/sor/db/astyanax/CqlDataWriterDAO.java

```diff
@@ -0,0 +1,211 @@
+        _updateMeter = metricRegistry.meter(getMetricName("updates"));
+        _oversizeUpdateMeter = metricRegistry.meter(getMetricName("oversizeUpdates"));
+        _deltaBlockSize = deltaBlockSize;
+        _deltaPrefix = deltaPrefix;
+        _deltaPrefixLength = deltaPrefix.length();
+    }
+
+    private String getMetricName(String name) {
+        return MetricRegistry.name("bv.emodb.sor", "CqlDataWriterDAO", name);
+    }
+
+    @Override
+    public long getFullConsistencyTimestamp(Table table) {
+        return _astyanaxWriterDAO.getFullConsistencyTimestamp(table);
+    }
+
+    @Override
+    public long getRawConsistencyTimestamp(Table table) {
+        return _astyanaxWriterDAO.getRawConsistencyTimestamp(table);
+    }
+
+    @Override
+    public void updateAll(Iterator<RecordUpdate> updates, UpdateListener listener) {
+        _astyanaxWriterDAO.updateAll(updates,listener);
+    }
+
+    @Override
+    public void compact(Table tbl, String key, UUID compactionKey, Compaction compaction, UUID changeId, Delta delta, Collection<UUID> changesToDelete, List<History> historyList, WriteConsistency consistency) {
+        checkNotNull(tbl, "table");
+        checkNotNull(key, "key");
+        checkNotNull(compactionKey, "compactionKey");
+        checkNotNull(compaction, "compaction");
+        checkNotNull(changeId, "changeId");
+        checkNotNull(delta, "delta");
+        checkNotNull(changesToDelete, "changesToDelete");
+        checkNotNull(consistency, "consistency");
+
+        AstyanaxTable table = (AstyanaxTable) tbl;
+        for (AstyanaxStorage storage : table.getWriteStorage()) {
+            DeltaPlacement placement = (DeltaPlacement) storage.getPlacement();
+            CassandraKeyspace keyspace = placement.getKeyspace();
+
+            ByteBuffer rowKey = storage.getRowKey(key);
+
+            // Should synchronously write compaction and then delete deltas
+            writeCompaction(rowKey, compactionKey, compaction, consistency, placement, keyspace, tbl, key);
+
+            deleteCompactedDeltas(rowKey, consistency, placement, keyspace, changesToDelete, historyList, tbl, key);
+        }
+    }
+
+    private void insertBlockedDeltas(BatchStatement batchStatement, DeltaTableDDL tableDDL, ConsistencyLevel consistencyLevel, ByteBuffer rowKey, UUID changeId, ByteBuffer encodedDelta, int deltaSize) {
+        int numBlocks = (deltaSize + _deltaBlockSize - 1) / _deltaBlockSize;
+        int position = encodedDelta.position();
+
+        byte[] blockBytes = String.format("%0" + _deltaPrefixLength + "X", numBlocks).getBytes();
+        for (int i = blockBytes.length - 1; i >= 0; i--) {
+            encodedDelta.put(position + _deltaPrefixLength - blockBytes.length + i, blockBytes[i]);
+        }
+        for (int block = 0; block < numBlocks; block++) {
+            ByteBuffer split = encodedDelta.duplicate();
+            int limit;
+            if (_deltaBlockSize * (block + 1) < deltaSize) {
```

Comment:
Notice that due to UTF-8 encoding the last block may turn out to be slightly larger on an edge condition with this implementation.  I personally don't think this is a huge deal, since each block would at most give up 3 bytes, meaning even with 1,000 blocks the final block would only be over the block size by 3k, and we could account for that by adding a small amount of padding to the CQL fetch size.  The alternative would be rewrite this loop so that it counts blocks as it creates them, then at the end goes back and adjusts the number of blocks written into the first block, currently taking place on lines 125-127.  That way each block is guaranteed to be no larger than the block size and bypasses the need to compute a possibly incorrect block count in adfance.